File: services/api/app/schemas/integration.py
"""
Integration schemas
"""

from typing import Optional, List, Dict, Any
from datetime import datetime
from pydantic import BaseModel, Field


class IntegrationPlatformBase(BaseModel):
    """Base integration platform schema"""

    id: str
    name: str
    description: str
    category: str
    logo: str
    coming_soon: bool = False
    permissions: List[str]
    modal_description: Optional[str] = None
    api_docs_url: Optional[str] = None
    display_order: int = 0
    is_active: bool = True


class IntegrationPlatformResponse(IntegrationPlatformBase):
    """Integration platform response"""

    class Config:
        from_attributes = True


class IntegrationPlatformList(BaseModel):
    """List of integration platforms"""

    platforms: List[IntegrationPlatformResponse]
    total: int


class UserIntegrationBase(BaseModel):
    """Base user integration schema"""

    platform_id: str


class UserIntegrationCreate(UserIntegrationBase):
    """Create user integration"""

    access_token: Optional[str] = None
    refresh_token: Optional[str] = None
    token_expires_at: Optional[datetime] = None
    scopes: Optional[List[str]] = None
    metadata: Optional[Dict[str, Any]] = None


class UserIntegrationUpdate(BaseModel):
    """Update user integration"""

    status: Optional[str] = None
    access_token: Optional[str] = None
    refresh_token: Optional[str] = None
    token_expires_at: Optional[datetime] = None
    last_sync_at: Optional[datetime] = None
    last_error: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None


class UserIntegrationResponse(BaseModel):
    """User integration response"""

    id: str
    user_id: str
    platform_id: str
    status: str
    connected_at: datetime
    last_sync_at: Optional[datetime] = None
    last_error: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None

    # Include platform info
    platform: Optional[IntegrationPlatformResponse] = None

    class Config:
        from_attributes = True


class ConnectedServiceResponse(BaseModel):
    """Connected service response for dashboard"""

    id: str
    provider_id: str
    name: str
    logo_url: str
    status: str  # active, needs_attention, expired
    connected_date: datetime
    last_sync_date: Optional[datetime]
    domain: str


class SyncStatusResponse(BaseModel):
    """Sync status response"""

    status: str  # success, in_progress, failed
    last_sync: Optional[datetime]
    domains: Dict[str, str]  # category -> status


class IntegrationOAuthStartRequest(BaseModel):
    """Request to start OAuth flow"""

    platform_id: str
    redirect_uri: str


class IntegrationOAuthStartResponse(BaseModel):
    """Response with OAuth URL"""

    authorization_url: str
    state: str


class IntegrationOAuthCallbackRequest(BaseModel):
    """OAuth callback data"""

    platform_id: str
    code: str
    state: str

File: apps/web/src/types/integration.ts
// types/integration.ts
export interface Provider {
  id: string;
  name: string;
  description: string;
  category: 'finance' | 'education' | 'career' | 'healthcare' | 'automotive' | 'smarthome';
  logo: string;
  connected: boolean;
  comingSoon?: boolean;
  permissions: string[];
  modalDescription?: string;
}

export interface ConnectedService {
  id: string;
  providerId: string;
  name: string;
  logoUrl: string;
  status: 'active' | 'needs_attention' | 'expired';
  connectedDate: string;
  lastSyncDate: string;
  domain: 'finance' | 'education' | 'career' | 'healthcare' | 'automotive' | 'smarthome';
  category?: string;
}

export interface IntegrationToken {
  id: string;
  providerId: string;
  accessToken: string;
  refreshToken?: string;
  expiresAt?: string;
  scopes: string[];
}

export interface SyncStatus {
  status: 'success' | 'in_progress' | 'failed';
  lastSync: string | null;
  domains: {
    finance: 'success' | 'in_progress' | 'failed';
    education: 'success' | 'in_progress' | 'failed';
    career: 'success' | 'in_progress' | 'failed';
    healthcare: 'success' | 'in_progress' | 'failed';
    automotive?: 'success' | 'in_progress' | 'failed';
    smarthome?: 'success' | 'in_progress' | 'failed';
  };
}

export interface SyncSource {
  provider: string;
  type: 'email' | 'calendar';
  status: string;
  lastSyncAt: string | null;
  recordCount: number;
}

export interface GraphRAGQueueStatus {
  pendingJobs: number;
  failedJobs: number;
  lastProcessedAt: string | null;
}

export interface SyncStatusResponse {
  sources: SyncSource[];
  graphrag: GraphRAGQueueStatus;
}

export interface SyncNowResponse {
  success: boolean;
  error?: string;
  retryAfterSeconds?: number;
  [key: string]: unknown;
}
